### PR TITLE
Add TypeConverter fallback to DefaultValueAttribute tests

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
closes https://github.com/dotnet/corefx/issues/30648

/cc @ericstj @Anipik @jkotas 

depends on https://github.com/dotnet/coreclr/pull/19354

### CODE REVIEW

**Comment by ericstj** [comment]
Can we also test the fallback if TypeConverter assembly is not present?  One way would be to use reflection to muck with internal state.  Another would be to set up some isolated test process and restrict its access to the TypeConverter assembly.  Not sure how easy the latter is to do.

**Comment by MarcoRossignoli** [comment]
> One way would be to use reflection to muck with internal state. Another would be to set up some isolated test process and restrict its access to the TypeConverter assembly. Not sure how easy the latter is to do.

We could mock with remote invoker(out of process to avoid interferences)...is there some place where i can see muck sample on codebase?

**Comment by ericstj** [comment]
If you're doing it out of process you might be able to prevent TC from ever loading with AssemblyLoadContext.  If you can do that, no need to muck with reflection.

Here's an example of test using reflection to try to get at internals and achieve coverage without running through the "actual" runtime scenario:
https://github.com/dotnet/corefx/blob/c263482ea552d02c6bb7f583374978be24d13cb5/src/Microsoft.Win32.SystemEvents/tests/SystemEventsTest.cs#L40
Here's another example:
https://github.com/dotnet/corefx/blob/39566c3c02061db78c6068ed8310a55d4757866c/src/System.DirectoryServices.Protocols/tests/SortRequestControlTests.cs#L100
Essentially the idea is to fake the state as if the load failed.  The challenge in this case since it is static state is you need to either isolate all tests that do this in their own app-domain/process, or restore the good state and ensure that no other tests are running when you have the bad state in place.  Given that I'd say you should go remote-invoker + assemblyloadcontext route.

**Comment by MarcoRossignoli** [comment]
@ericstj thank's!I'll work on it and get back to you!

**Comment by MarcoRossignoli** [comment]
@ericstj i'm working on Type.GetType mock. After some check seems that call go to [internal](https://github.com/dotnet/coreclr/blob/4f92fa9f95b0e45935374da29066978719a7701d/src/System.Private.CoreLib/src/System/RuntimeHandles.cs#L427). 
What do you think of idea to add a `static string typeDescriptorAQN = "System.ComponentModel.TypeDescriptor, System, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089";` on DefaultValueAttribute that i can "change" via reflection to a "inexistent" type to return null from GetType call?

/cc @jkotas

**Comment by jkotas** [comment]
If you want to use private reflection, why not just change `s_convertFromInvariantString` that you have there to `new object()` to test the other path?

**Comment by MarcoRossignoli** [comment]
> If you want to use private reflection, why not just change s_convertFromInvariantString that you have there to new object() to test the other path?

 I thought about it, but to "cover" much code as possible my idea was to "simulate" `null` return. Do you think is enough to only cover fallbacks?

**Comment by jkotas** [comment]
If you wanted to really simulate the missing `System.ComponentModel.TypeDescriptor`, you should do it using AssemblyLoadContext like Eric mentioned. 

Here is a small standalone example - it creates custom AssemblyLoadContext and blocks loading System.ComponentModel.TypeConverter into it. If the hello.dll tries to Load it, it will behave as it was not there:

```
using System;
using System.IO;
using System.Reflection;
using System.Runtime.Loader;

class MyAssemblyLoadContext : AssemblyLoadContext
{
    protected override Assembly Load(AssemblyName name)
    {
        if (name.Name == "System.ComponentModel.TypeConverter")
            throw new FileNotFoundException();
        return null;
    }
}

class Program
{
    static void Main(string[] args)
    {
        var alc = new MyAssemblyLoadContext();
        var asm = alc.LoadFromAssemblyPath(@"C:\repro\hello\obj\Debug\netcoreapp2.1\hello.dll");
        asm.EntryPoint.Invoke(null, new object[] { new string[0] });
    }
}
```

**Comment by ericstj** [comment]
@jkotas, my impression was that wouldn't work since core lib was doing the load and you cannot control the AssemblyLoadContext used by core lib.

**Comment by MarcoRossignoli** [comment]
> you cannot control the AssemblyLoadContext used by core lib

mmm...some clarification...are you saying that on corelib there is an "private instance"of AssemblyLoadContext?

**Comment by jkotas** [comment]
> Type is loaded in "Default LoadContext "and we cannot interfere, right?

Right.

I think setting the static using reflection to `new object()` has the right tradeoff, given the different options.

**Comment by MarcoRossignoli** [comment]
> I think setting the static using reflection to new object() has the right tradeoff, given the different options.

Perfect! Thank's to all!

**Comment by MarcoRossignoli** [comment]
@ericstj @jkotas  added fallback tests.

**Comment by MarcoRossignoli** [comment]
@dotnet-bot build this please

**Comment by MarcoRossignoli** [comment]
@dotnet-bot test this please

**Comment by MarcoRossignoli** [comment]
@ericstj @jkotas let me know if there are other works to di here for merge!

File: src/System.Runtime/tests/System/ComponentModel/DefaultValueAttributeTests.cs [hunk 1]
```diff
@@ -35,6 +38,88 @@ public static void Ctor()
...
+        [Fact]
+        public static void Ctor_CustomTypeConverter()
+        {
+            TypeDescriptor.AddAttributes(typeof(CustomType), new TypeConverterAttribute(typeof(CustomConverter)));
+            DefaultValueAttribute attr = new DefaultValueAttribute(typeof(CustomType), "42");
+            Assert.Equal(42, ((CustomType)attr.Value).Value);
+        }
+
+        [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
```
Comment (Anipik) [inline src/System.Runtime/tests/System/ComponentModel/DefaultValueAttributeTests.cs]: can you add a reason why we are skipping on .net framework

File: src/System.Runtime/tests/System/ComponentModel/DefaultValueAttributeTests.cs [hunk 2]
```diff
@@ -35,6 +38,88 @@ public static void Ctor()
...
+
+                // simulate TypeDescriptor.ConvertFromInvariantString not found
+                s_convertFromInvariantString.SetValue(null, new object());
+
+                // we fallback to Convert.ChangeType in DefaultValueAttribute constructor
+                DefaultValueAttribute attr = new DefaultValueAttribute(typeof(int), "42");
+
+                Assert.Equal(42, attr.Value);
+
+                return (int)attr.Value;
```
Comment (Anipik) [inline src/System.Runtime/tests/System/ComponentModel/DefaultValueAttributeTests.cs]: why we are returning the value here ? its not being used anywhere

File: src/System.Runtime/tests/System/ComponentModel/DefaultValueAttributeTests.cs [hunk 3]
```diff
@@ -35,6 +38,88 @@ public static void Ctor()
...
+
+        [Fact]
+        public static void Ctor_CustomTypeConverter_Null()
+        {
+            DefaultValueAttribute attr = new DefaultValueAttribute(typeof(CustomType2), "42");
+            Assert.Null(attr.Value);
+        }
+
+        [Fact]
+        public static void Ctor_DefaultTypeConverter_Null()
```
Comment (Anipik) [inline src/System.Runtime/tests/System/ComponentModel/DefaultValueAttributeTests.cs]: Combine these two tests using [theory]

**Comment by MarcoRossignoli** [comment]
@ericstj do i need other work to merge?

File: src/System.Runtime/tests/System/ComponentModel/DefaultValueAttributeTests.cs [hunk 4]
```diff
@@ -35,6 +38,81 @@ public static void Ctor()
...
+        {
+            TypeDescriptor.AddAttributes(typeof(CustomType), new TypeConverterAttribute(typeof(CustomConverter)));
+            DefaultValueAttribute attr = new DefaultValueAttribute(typeof(CustomType), "42");
+            Assert.Equal(42, ((CustomType)attr.Value).Value);
+        }
+
+        [Fact]
+        // On NetFramework will fail because there isn't fallback code, only call to TypeDescriptor.GetConverter
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
+        public void Ctor_CustomTypeConverter_TypeDescriptorNotFound_ExceptionFallback()
```
[the remark this answers is not in the document]
Comment (MarcoRossignoli) [reply src/System.Runtime/tests/System/ComponentModel/DefaultValueAttributeTests.cs]: ok wait for merge i'll fix it.

[the remark this answers is not in the document]
Comment (MarcoRossignoli) [reply src/System.Runtime/tests/System/ComponentModel/DefaultValueAttributeTests.cs]: I thought to put types test togheter, but with separate tests i'm sure that we have clean runtime envinroment.

**Pull request merged**